lib/TokenManager.js

```diff
@@ -141,29 +139,27 @@ function renew(sdk, tokenMgmtRef, storage, key) {
   if (!tokenMgmtRef.renewPromise[key]) {
     tokenMgmtRef.renewPromise[key] = sdk.token.renew(token)
     .then(function(freshTokens) {
-      // We may receive more tokens than we requested
-      var map = {};
-      if (freshTokens instanceof Array === false) {
-        freshTokens = [freshTokens];
+      var freshToken = freshTokens;
+      // With PKCE flow we will receive multiple tokens. Find the one we are looking for
+      if (freshTokens instanceof Array) {
+        freshToken = freshTokens.filter(function(freshToken) {
+          return (freshToken.idToken && token.idToken) || (freshToken.accessToken && token.accessToken);
+        }).pop();
       }
-      freshTokens.forEach(function(freshToken) {
-        var inferredKey = freshToken.idToken ? 'idToken' : freshToken.accessToken ? 'accessToken' : key;
-        map[inferredKey] = freshToken;
-
-        var oldToken = get(storage, inferredKey);
-        if (!oldToken) {
-          // It is possible to enter a state where the tokens have been cleared
-          // after a renewal request was triggered. To ensure we do not store a
-          // renewed token, we verify the promise key doesn't exist and return.
-          return;
-        }
-        add(sdk, tokenMgmtRef, storage, inferredKey, freshToken);
-        tokenMgmtRef.emitter.emit('renewed', inferredKey, freshToken, oldToken);
-      });
 
+      var oldToken = get(storage, key);
+      if (!oldToken) {
```

Comment:
Since this is an early exit, why not do that BEFORE looking at freshToken?

Reply:
It is checked, at the beginning of the function. But there was some XHR since then, this is an additional check during the response. I believe this is to catch an async error where tokens are cleared while renew is in-flight. (I did not write this, it existed before the PKCE refactor)